File: java/src/main/java/io/yosina/transliterators/HiraKataTransliterator.java
package io.yosina.transliterators;

import io.yosina.Char;
import io.yosina.CharIterator;
import io.yosina.CodePointTuple;
import io.yosina.Transliterator;
import io.yosina.annotations.RegisteredTransliterator;
import java.util.HashMap;
import java.util.Map;

/** Transliterator for converting between Hiragana and Katakana scripts. */
@RegisteredTransliterator(name = "hira-kata")
public class HiraKataTransliterator implements Transliterator {
    /** Options for configuring the behavior of HiraKataTransliterator. */
    public static class Options {
        private final Mode mode;

        /** Conversion mode. */
        public enum Mode {
            /** Convert Hiragana to Katakana */
            HIRA_TO_KATA,
            /** Convert Katakana to Hiragana */
            KATA_TO_HIRA
        }

        /**
         * Creates a new Options instance with the specified mode.
         *
         * @param mode the conversion mode
         */
        public Options(Mode mode) {
            this.mode = mode;
        }

        /**
         * Gets the conversion mode.
         *
         * @return the conversion mode
         */
        public Mode getMode() {
            return mode;
        }
    }

    // Class-level cache for mapping tables (package-private for testing)
    static final Map<Options.Mode, Map<Integer, Integer>> mappingCache = new HashMap<>();

    private final Map<Integer, Integer> mappingTable;

    /**
     * Creates a new HiraKataTransliterator with the specified options.
     *
     * @param options the transliterator options
     */
    public HiraKataTransliterator(Options options) {
        this.mappingTable = getMappingTable(options.getMode());
    }

    /** Default constructor uses hira-to-kata mode. */
    public HiraKataTransliterator() {
        this(new Options(Options.Mode.HIRA_TO_KATA));
    }

    private static synchronized Map<Integer, Integer> getMappingTable(Options.Mode mode) {
        // Check cache first
        Map<Integer, Integer> cached = mappingCache.get(mode);
        if (cached != null) {
            return cached;
        }

        Map<Integer, Integer> mapping = new HashMap<>();

        // Main table mappings
        for (HiraKataTable.HiraKataEntry entry : HiraKataTable.HIRAGANA_KATAKANA_TABLE) {
            if (entry.hiragana.isPresent()) {
                HiraKataTable.HiraKata hira = entry.hiragana.get();
                HiraKataTable.HiraKata kata = entry.katakana;

                if (mode == Options.Mode.HIRA_TO_KATA) {
                    mapping.put(hira.base, kata.base);
                    if (hira.voiced >= 0 && kata.voiced >= 0) {
                        mapping.put(hira.voiced, kata.voiced);
                    }
                    if (hira.semivoiced >= 0 && kata.semivoiced >= 0) {
                        mapping.put(hira.semivoiced, kata.semivoiced);
                    }
                } else {
                    mapping.put(kata.base, hira.base);
                    if (kata.voiced >= 0 && hira.voiced >= 0) {
                        mapping.put(kata.voiced, hira.voiced);
                    }
                    if (kata.semivoiced >= 0 && hira.semivoiced >= 0) {
                        mapping.put(kata.semivoiced, hira.semivoiced);
                    }
                }
            }
        }

        // Small character mappings
        for (HiraKataTable.SmallKanaEntry entry : HiraKataTable.HIRAGANA_KATAKANA_SMALL_TABLE) {
            if (mode == Options.Mode.HIRA_TO_KATA) {
                mapping.put(entry.hiragana, entry.katakana);
            } else {
                mapping.put(entry.katakana, entry.hiragana);
            }
        }

        // Cache the result
        mappingCache.put(mode, mapping);
        return mapping;
    }

    @Override
    public CharIterator transliterate(CharIterator inputChars) {
        return new CharIterator() {
            @Override
            public boolean hasNext() {
                return inputChars.hasNext();
            }

            @Override
            public Char next() {
                Char ch = inputChars.next();
                CodePointTuple tuple = ch.get();
                if (tuple.size() == 1) {
                    int codePoint = tuple.get(0);
                    Integer mapped = mappingTable.get(codePoint);

                    if (mapped != null) {
                        return new Char(CodePointTuple.of(mapped), ch.getOffset(), ch);
                    }
                }
                return ch;
            }
        };
    }
}

File: csharp/src/Yosina/Transliterators/HiraKataTransliterator.cs
// Copyright (c) Yosina. All rights reserved.

using System.Collections.Concurrent;
using System.Text.Json.Serialization;
using Yosina.JsonConverters;

namespace Yosina.Transliterators;

/// <summary>
/// Transliterator for converting between Hiragana and Katakana scripts.
/// </summary>
[RegisteredTransliterator("hira-kata")]
public class HiraKataTransliterator : ITransliterator
{
    /// <summary>
    /// Conversion mode.
    /// </summary>
    [JsonConverter(typeof(JsonEnumValueConverter<Mode>))]
    public enum Mode
    {
        /// <summary>
        /// Convert Hiragana to Katakana.
        /// </summary>
        [JsonEnumValue("hira-to-kata")]
        HiraToKata,

        /// <summary>
        /// Convert Katakana to Hiragana.
        /// </summary>
        [JsonEnumValue("kata-to-hira")]
        KataToHira,
    }

    /// <summary>
    /// Options for configuring the behavior of HiraKataTransliterator.
    /// </summary>
    public class Options
    {
        /// <summary>
        /// Gets or sets the conversion mode.
        /// </summary>
        [JsonPropertyName("mode")]
        public Mode Mode { get; set; } = Mode.HiraToKata;
    }

    // Class-level cache for mapping tables
    private static readonly ConcurrentDictionary<Mode, Dictionary<int, int>> MappingCache = new ConcurrentDictionary<Mode, Dictionary<int, int>>();

    private readonly Dictionary<int, int> mappingTable;

    /// <summary>
    /// Initializes a new instance of the <see cref="HiraKataTransliterator"/> class with the specified options.
    /// </summary>
    /// <param name="options">The options for the transliterator.</param>
    public HiraKataTransliterator(Options options)
    {
        this.mappingTable = GetMappingTable(options.Mode);
    }

    /// <summary>
    /// Initializes a new instance of the <see cref="HiraKataTransliterator"/> class with default options.
    /// </summary>
    public HiraKataTransliterator()
        : this(new Options())
    {
    }

    private static Dictionary<int, int> GetMappingTable(Mode mode)
    {
        // Check cache first
        if (MappingCache.TryGetValue(mode, out var cached))
        {
            return cached;
        }

        var mapping = new Dictionary<int, int>();

        // Main table mappings
        foreach (var entry in HiraKataTable.HiraganaKatakanaTable)
        {
            if (entry.Hiragana.HasValue)
            {
                var hira = entry.Hiragana.Value;
                var kata = entry.Katakana;

                if (mode == Mode.HiraToKata)
                {
                    mapping[hira.Base] = kata.Base;
                    if (hira.Voiced >= 0 && kata.Voiced >= 0)
                    {
                        mapping[hira.Voiced] = kata.Voiced;
                    }

                    if (hira.Semivoiced >= 0 && kata.Semivoiced >= 0)
                    {
                        mapping[hira.Semivoiced] = kata.Semivoiced;
                    }
                }
                else
                {
                    mapping[kata.Base] = hira.Base;
                    if (kata.Voiced >= 0 && hira.Voiced >= 0)
                    {
                        mapping[kata.Voiced] = hira.Voiced;
                    }

                    if (kata.Semivoiced >= 0 && hira.Semivoiced >= 0)
                    {
                        mapping[kata.Semivoiced] = hira.Semivoiced;
                    }
                }
            }
        }

        // Small character mappings
        foreach (var entry in HiraKataTable.HiraganaKatakanaSmallTable)
        {
            if (mode == Mode.HiraToKata)
            {
                mapping[entry.Hiragana] = entry.Katakana;
            }
            else
            {
                mapping[entry.Katakana] = entry.Hiragana;
            }
        }

        // Cache the result
        MappingCache[mode] = mapping;
        return mapping;
    }

    /// <inheritdoc/>
    public IEnumerable<Character> Transliterate(IEnumerable<Character> input)
    {
        foreach (var ch in input)
        {
            if (ch.CodePoint.Size() == 1)
            {
                var codePoint = ch.CodePoint.First;
                if (this.mappingTable.TryGetValue(codePoint, out var mapped))
                {
                    yield return new Character((mapped, CodePointTupleExtensions.INVALID), ch.Offset, ch);
                    continue;
                }
            }

            yield return ch;
        }
    }
}
